core/src/core/merkle_proof.rs

```diff
@@ -0,0 +1,159 @@
+	/// The size of the MMR at the time the proof was created.
+	pub mmr_size: u64,
+	/// The sibling path from the leaf up to the final sibling hashing to the
+	/// root.
+	pub path: Vec<Hash>,
+}
+
+impl Writeable for MerkleProof {
+	fn write<W: Writer>(&self, writer: &mut W) -> Result<(), ser::Error> {
+		writer.write_u64(self.mmr_size)?;
+		writer.write_u64(self.path.len() as u64)?;
+		self.path.write(writer)?;
+		Ok(())
+	}
+}
+
+impl Readable for MerkleProof {
+	fn read(reader: &mut Reader) -> Result<MerkleProof, ser::Error> {
+		let mmr_size = reader.read_u64()?;
+		let path_len = reader.read_u64()?;
+		let mut path = Vec::with_capacity(path_len as usize);
+		for _ in 0..path_len {
+			let hash = Hash::read(reader)?;
+			path.push(hash);
+		}
+
+		Ok(MerkleProof { mmr_size, path })
+	}
+}
+
+impl Default for MerkleProof {
+	fn default() -> MerkleProof {
+		MerkleProof::empty()
+	}
+}
+
+impl MerkleProof {
+	/// The "empty" Merkle proof.
+	pub fn empty() -> MerkleProof {
+		MerkleProof {
+			mmr_size: 0,
+			path: Vec::default(),
+		}
+	}
+
+	/// Serialize the Merkle proof as a hex string (for api json endpoints)
+	pub fn to_hex(&self) -> String {
+		let mut vec = Vec::new();
+		ser::serialize(&mut vec, &self).expect("serialization failed");
+		util::to_hex(vec)
+	}
+
+	/// Convert hex string represenation back to a Merkle proof instance
+	pub fn from_hex(hex: &str) -> Result<MerkleProof, String> {
+		let bytes = util::from_hex(hex.to_string()).unwrap();
+		let res = ser::deserialize(&mut &bytes[..])
+			.map_err(|_| format!("failed to deserialize a Merkle Proof"))?;
+		Ok(res)
+	}
+
+	/// Verifies the Merkle proof against the provided
+	/// root hash, element and position in the MMR.
+	pub fn verify(
+		&self,
+		root: Hash,
+		element: &PMMRIndexHashable,
+		node_pos: u64,
+	) -> Result<(), MerkleProofError> {
+		let mut proof = self.clone();
+		proof.verify_consume(root, element, node_pos)
+	}
+
+	/// Consumes the Merkle proof while verifying it.
+	/// The proof can no longer beused by the caller after dong this.
+	/// Caller must clone() the proof first.
+	fn verify_consume(
+		&mut self,
+		root: Hash,
+		element: &PMMRIndexHashable,
+		node_pos: u64,
+	) -> Result<(), MerkleProofError> {
+		let node_hash = if node_pos > self.mmr_size {
+			element.hash_with_index(self.mmr_size)
+		} else {
+			element.hash_with_index(node_pos - 1)
+		};
+
+		// handle special case of only a single entry in the MMR
+		// (no siblings to hash together)
+		if self.path.len() == 0 {
+			if root == node_hash {
+				return Ok(());
+			} else {
+				return Err(MerkleProofError::RootMismatch);
+			}
+		} else if self.mmr_size == 1 {
+			if self.path.len() == 1 && root == node_hash && vec![root] == self.path {
+				return Ok(());
+			} else {
+				return Err(MerkleProofError::RootMismatch);
+			}
+		}
+
+		let sibling = self.path.remove(0);
+		let (parent_pos, sibling_pos) = pmmr::family(node_pos);
+
+		let peaks_pos = pmmr::peaks(self.mmr_size);
```

Comment:
determining peaks at every step of a merkle path is rather inefficient, as are all the family calls. the computed node positions above the peaks also don't make much sense, but it does look like it should produce the right result. i plan to rewrite all this in the near future anyway:-)